mono/mini/mini-llvm.c

```diff
@@ -3968,6 +3992,45 @@ process_call (EmitContext *ctx, MonoBasicBlock *bb, LLVMBuilderRef *builder_ref,
 	 */
 	lcall = emit_call (ctx, bb, &builder, callee, args, LLVMCountParamTypes (llvm_sig));
 
+	if (mono_aot_can_specialize (call->method)) {
+		LLVMValueRef lmethod = callee;
+
+		int num_params = LLVMCountParams (lmethod);
+		GArray *call_site_union = (GArray *) g_hash_table_lookup (ctx->module->method_to_call_info, call->method);
+
+		if (!call_site_union) {
+			call_site_union = g_array_sized_new (FALSE, TRUE, sizeof (gint32), num_params);
+			int zero = 0;
+			for (int i = 0; i < num_params; i++)
+				g_array_insert_val (call_site_union, i, zero);
+		}
+
+		for (int i = 0; i < num_params; i++) {
+			LLVMValueRef argument = args [i];
+
+			// Add to list of tracked callers/callees
+			GList *usages = (GList *) g_hash_table_lookup (ctx->module->lvalue_to_lcalls, argument);
+			usages = g_list_prepend (usages, lcall);
+			g_hash_table_insert (ctx->module->lvalue_to_lcalls, argument, usages);
```

Comment:
I'm not convinced this line does anything if there are 2 or more uses of an argument.  `g_hash_table_insert` (as opposed to `g_hash_table_replace`) is a no-op if `argument` was already in the table.   And it clearly must have been because we just looked it up.

Reply:
The official docs seem to suggest otherwise. 

https://developer.gnome.org/glib/stable/glib-Hash-Tables.html

```
g_hash_table_insert (GHashTable *hash_table,
                     gpointer key,
                     gpointer value);
```

```
If the key already exists in the GHashTable its *current value is replaced with the new value*. If you supplied a value_destroy_func when creating the GHashTable, the old value is freed using that function. If you supplied a key_destroy_func when creating the GHashTable, the passed key is freed using that function.
```

Does our eglib differ from glib here?

Reply:
ah interesting.
```
#define g_hash_table_insert(h,k,v)    g_hash_table_insert_replace ((h),(k),(v),FALSE)
#define g_hash_table_replace(h,k,v)   g_hash_table_insert_replace ((h),(k),(v),TRUE)
```
so that last `gboolean replace` parameter is about replacing the _key_.  The value is always replaced.   this is controlling whether to replace the key in the case when you have a custom key equivalence function.